server/core/src/password_hash.cpp

```diff
@@ -0,0 +1,311 @@
+#include "irods/irods_random.hpp"
+#include "irods/irods_exception.hpp"
+#include "irods/rodsErrorTable.h"
+#include "irods/irods_logger.hpp"
+
+#include <fmt/format.h>
+#include <nlohmann/json.hpp>
+
+#include <openssl/core_names.h>
+#include <openssl/kdf.h>
+#include <openssl/params.h>
+
+#include <algorithm>
+#include <array>
+#include <string>
+#include <vector>
+
+namespace
+{
+    using log_server = irods::experimental::log::server;
+
+    class key_derivation_function
+    {
+      public:
+        key_derivation_function() = default;
+
+        // Move constructor and assignment - okay.
+        key_derivation_function(key_derivation_function&& other) = default;
+        key_derivation_function& operator=(key_derivation_function&& other) = default;
+
+        // Copy constructor and assignment - not allowed.
+        key_derivation_function(key_derivation_function& other) = delete;
+        key_derivation_function& operator=(key_derivation_function& other) = delete;
+
+        virtual ~key_derivation_function() = default;
+
+        virtual auto derive_key(const std::string& _password, const std::string& _salt)
+            -> std::vector<unsigned char> = 0;
+    }; // class key_derivation_function
+
+    class scrypt_kdf : public key_derivation_function
+    {
+      public:
+        explicit scrypt_kdf(const nlohmann::json& _parameters)
+            : kctx{}
+            , work_factor{}
+            , resources{}
+            , parallelism{}
+            , keylen{}
+            , algorithm_params{}
+        {
+            log_server::trace("params: {}", _parameters.dump());
+
+            // Get the desired key length (dkLen).
+            const auto keylen_iter = _parameters.find("key_length");
+            if (keylen_iter == _parameters.end()) {
+                THROW(SYS_INVALID_INPUT_PARAM, fmt::format("{}: key_length missing from scrypt parameters.", __func__));
+            }
+            const auto keylen_input = keylen_iter->get<std::int32_t>();
+            // Even though there is no practical maximum to the desired key length, a maximum of 1 GiB is decided
+            // arbitrarily here because it is so large as to be impractical as a password hashing key length. The KDF
+            // derive function allows for larger values than this, but this was done so that underflows are detected.
+            constexpr std::int32_t maximum_key_length_in_bytes = 1073741824;
+            if (keylen_input < 1 || keylen_input > maximum_key_length_in_bytes) {
+                THROW(SYS_INVALID_INPUT_PARAM, fmt::format("{}: key_length must be a positive integer.", __func__));
+            }
+            keylen = static_cast<std::uint32_t>(keylen_input);
+
+            // Get the cost factor parameter (N).
+            const auto n_iter = _parameters.find("N");
+            if (n_iter == _parameters.end()) {
+                THROW(SYS_INVALID_INPUT_PARAM, fmt::format("{}: N missing from scrypt parameters.", __func__));
```

Comment:
Might want to qualify the N, r, and p with a trailing "parameter".

e.g. `N parameter missing from ...`